collection/range-set.rkt

```diff
@@ -0,0 +1,332 @@
+#lang racket/base
+
+
+(require racket/contract/base)
+
+
+(provide
+ (contract-out
+  [range-set? predicate/c]
+  [range-set (-> nonempty-range? ... range-set?)]
+  [range-set-size (-> range-set? natural?)]
+  [sequence->range-set (-> (sequence/c nonempty-range?) range-set?)]
+  [into-range-set (reducer/c nonempty-range? range-set?)]
+  [in-range-set (-> range-set? (sequence/c nonempty-range?))]
+  [empty-range-set empty-range-set?]
+  [empty-range-set? predicate/c]
+  [nonempty-range-set? predicate/c]
+  [range-set-contains? (-> range-set? any/c boolean?)]
+  [range-set-encloses? (-> range-set? range? boolean?)]
+  [range-set-encloses-all? (-> range-set? (or/c range-set? (sequence/c range?)) boolean?)]
+  [range-subset (-> range-set? range? range-set?)]))
+
+
+(require racket/match
+         racket/math
+         racket/sequence
+         racket/struct
+         racket/vector
+         rebellion/base/comparator
+         rebellion/base/option
+         rebellion/base/range
+         (submod rebellion/base/range private-for-range-set-implementation-only)
+         rebellion/collection/vector/builder
+         rebellion/private/guarded-block
+         rebellion/private/static-name
+         rebellion/streaming/reducer
+         rebellion/streaming/transducer
+         rebellion/type/tuple)
+
+
+(module+ test
+  (require (submod "..")
+           rackunit))
+
+
+;@----------------------------------------------------------------------------------------------------
+;; Vector utilities
+
+
+;; Vector A, BiPredicate A, BinaryOperator A -> ImmutableVector A
+;; Returns a new vector that is like vec, except adjacent elements are merged with merge-function when
+;; should-merge? returns true.
+;; Examples:
+;; > (vector-merge-adjacent (vector 1 2 3 "hello" 4 5 "world" 6) both-numbers? +)
+;; (vector 6 "hello" 9 "world" 6)
+(define (vector-merge-adjacent vec should-merge? merge-function)
+  (define count (vector-length vec))
+  (define/guard (iteration-loop builder i)
+    (guard (equal? i count) then
+      (build-vector builder))
+    (define/guard (merge-loop element j)
+      (guard (equal? j count) then
+        (build-vector (vector-builder-add builder element)))
+      (define next-element (vector-ref vec j))
+      (if (should-merge? element next-element)
+          (merge-loop (merge-function element next-element) (add1 j))
+          (iteration-loop (vector-builder-add builder element) j)))
+    (merge-loop (vector-ref vec i) (add1 i)))
+  (iteration-loop (make-vector-builder #:expected-size count) 0))
```

Comment:
We can use a single `for/fold` here to avoid the both confusing nested iteration structure as well as all the index manipulation:
 
```suggestion
  (guard (< count 2) then
         (vector->immutable-vector vec))
  (for/fold ([builder (make-vector-builder #:expected-size count)]
             [element (vector-ref vec 0)]
             #:result (build-vector (vector-builder-add builder element)))
            ([next-element (in-vector vec 1)])
    (if (should-merge? element next-element)
        (values builder
                (merge-function element next-element))
        (values (vector-builder-add builder element)
                next-element))))
```